// mofron-comp-menutext/index.js
/**
 * @file mofron-comp-menutext/index.js
 * @brief menu text component
 * @license MIT
 */
const Text = require("mofron-comp-text");
const Click = require("mofron-event-click");
const loMargin = require("mofron-layout-margin");
const ConfArg = mofron.class.ConfArg;

module.exports = class extends mofron.class.Component {
    /**
     * initialize component
     * 
     * @param (mixed) 
     *                key-value: component config
     * @short 
     * @type private
     */
    constructor (p1) {
        try {
            super();
            this.modname("MenuText");
	    this.shortForm("text");
            
	    /* init config */
            this.confmng().add("selectEvent", { type:'event', list:true  })
            
	    if (0 < arguments.length) {
                this.config(p1);
            }
        } catch (e) {
            console.error(e.stack);
            throw e;
        }
    }
    
    /**
     * initialize dom contents
     * 
     * @type private
     */
    initDomConts () {
        try {
            super.initDomConts();
	    this.layout(new loMargin("top","0.1rem"));
        } catch (e) {
            console.error(e.stack);
            throw e;
        }
    }

    beforeRender () {
        try {
            super.beforeRender();
            this.select(0);
	} catch (e) {
            console.error(e.stack);
            throw e;
        }
    }



    text (prm) {
        try {
	    this.child(
	        new Text({
		    text: prm,
		    size:'0.18rem',
		    event:new Click(new ConfArg(this.clickEvent,this))
		})
            );
	} catch (e) {
            console.error(e.stack);
            throw e;
        }
    }

    clickEvent (p1,p2,p3) {
        try {
	    p1.id()
            let chd = p3.child();
            for (let cidx in chd) {
                if (chd[cidx].id() == p1.id()) {
                    p3.select(cidx);
                    break;
		}
	    }
	} catch (e) {
            console.error(e.stack);
            throw e;
        }
    }
    
    select (idx) {
        try {
            let chd = this.child();
            chd.forEach((c, cidx) => {
                c.weight(300);
            });
            chd[idx].weight(700);
            
            let menu    = this;
	    let sel_evt = this.selectEvent();
            sel_evt.forEach((itm) => {
                itm[0](menu.child()[idx], idx, itm[1]);
	    });
	} catch (e) {
            console.error(e.stack);
            throw e;
        }
    }
    
    selectEvent (fnc, prm) {
        try {
            if (undefined === fnc) {
                return this.confmng("selectEvent");
            }
            this.confmng("selectEvent", [fnc,prm]);
	} catch (e) {
            console.error(e.stack);
            throw e;
        }
    }
    
}
/* end of file */
